src/promptflow/promptflow/azure/_storage/cosmosdb/span.py

```diff
@@ -42,21 +47,40 @@ def __init__(self, span: SpanEntity, collection_id: str, created_by: dict) -> No
         self.collection_id = collection_id
         self.id = span.span_id
         self.created_by = created_by
+        self.external_event_data_uris = []
 
-    def persist(self, client):
+    def persist(self, cosmos_client: ContainerProxy, blob_container_client: ContainerClient, blob_base_uri: str):
         if self.id is None or self.partition_key is None or self.resource is None:
             return
 
+        if self.events and blob_container_client is not None and blob_base_uri is not None:
+            self._persist_events(blob_container_client, blob_base_uri)
+
         resource_attributes = self.resource.get(SpanFieldName.ATTRIBUTES, None)
         if resource_attributes is None:
             return
 
         from azure.cosmos.exceptions import CosmosResourceExistsError
 
         try:
-            return client.create_item(body=self.to_dict())
+            return cosmos_client.create_item(body=self.to_dict())
         except CosmosResourceExistsError:
             return None
 
     def to_dict(self) -> Dict[str, Any]:
         return {k: v for k, v in self.__dict__.items() if v}
+
+    def _persist_events(self, blob_container_client: ContainerClient, blob_base_uri: str):
+        idx = 0
+        for event in self.events:
+            event_data = json.dumps(event)
+            blob_client = blob_container_client.get_blob_client(self._event_path(idx))
+            blob_client.upload_blob(event_data)
+
+            event[SpanEventFieldName.ATTRIBUTES] = {}
+            self.external_event_data_uris.append(f"{blob_base_uri}{self._event_path(idx)}")
+
+            idx += 1
+
+    def _event_path(self, idx: int) -> str:
+        return f".promptflow/.trace/{self.collection_id}/{self.context[SpanContextFieldName.TRACE_ID]}/{self.id}/{idx}"
```

Comment:
How about extract the hard code part as constant for maintainability.

```python
EVENT_PATH_PREFIX = ".promptflow/.trace/"

def _event_path(self, idx: int) -> str:
    return f"{EVENT_PATH_PREFIX}{self.collection_id}/{self.context[SpanContextFieldName.TRACE_ID]}/{self.id}/{idx}"
```

Reply:
Fixed. Thanks.